source/MaterialXGenShader/UnitConverter.h

```diff
@@ -0,0 +1,218 @@
+    {
+        return _unitType;
+    }
+
+    /// Return the name of the default unit
+    const string& getDefaultUnit() const
+    {
+        return _defaultUnit;
+    }
+
+    /// @name Conversion
+    /// @{
+
+    /// Return the mappings from unit names to the scale value
+    /// defined by a linear converter. 
+    const std::unordered_map<string, float>& getUnitScale() const
+    {
+        return _unitScale;
+    }
+
+    /// Ratio between the given unit to a desired unit
+    /// @param inputUnit Unit of input value
+    /// @param outputUnit Unit for output value
+    float conversionRatio(const string& inputUnit, const string& outputUnit) const;
+
+    /// Convert a given value in a given unit to a desired unit
+    /// @param input Input value to convert
+    /// @param inputUnit Unit of input value
+    /// @param outputUnit Unit for output value
+    float convert(float input, const string& inputUnit, const string& outputUnit) const override;
+
+    /// Convert a given value in a given unit to a desired unit
+    /// @param input Input value to convert
+    /// @param inputUnit Unit of input value
+    /// @param outputUnit Unit for output value
+    Vector2 convert(Vector2 input, const string& inputUnit, const string& outputUnit) const override;
+
+    /// Convert a given value in a given unit to a desired unit
+    /// @param input Input value to convert
+    /// @param inputUnit Unit of input value
+    /// @param outputUnit Unit for output value
+    Vector3 convert(Vector3 input, const string& inputUnit, const string& outputUnit) const override;
+
+    /// Convert a given value in a given unit to a desired unit
+    /// @param input Input value to convert
+    /// @param inputUnit Unit of input value
+    /// @param outputUnit Unit for output value
+    Vector4 convert(Vector4 input, const string& inputUnit, const string& outputUnit) const override;
+
+    /// @}
+    /// @name Shader Mapping
+    /// @{
+
+    /// Given a unit name return a value that it can map to as an integer.
+    /// Returns -1 value if not found
+    int getUnitAsInteger(const string& unitName) const override;
+
+    /// Given an integer index return the unit name in the map used by the converter.
+    /// Returns Empty string if not found
+    virtual string getUnitFromInteger(int index) const override;
+
+    /// @}
+
+  private:
+    DefaultUnitConverter(UnitTypeDefPtr UnitDef);
+
+    std::unordered_map<string, float> _unitScale;
+    std::unordered_map<string, int> _unitEnumeration;
+    string _defaultUnit;
+    string _unitType;
+};
+
+
+class UnitConverterRegistry;
+
+/// A shared pointer to an UnitConverterRegistry
+using UnitConverterRegistryPtr = shared_ptr<UnitConverterRegistry>;
+/// A shared pointer to a const UnitConverterRegistry
+using ConstUnitConverterRegistryPtr = shared_ptr<const UnitConverterRegistry>;
+
+/// @class UnitConverterRegistry
+/// A registry of unit converters.
+///
+class UnitConverterRegistry
```

Comment:
What is the conceptual distinction between UnitConverterRegistry and UnitSystem?  Could the former be merged into the latter, making it responsible for managing the list of converters as well?

One related note is that we might consider the name UnitManagementSystem instead of UnitSystem, making a clearer parallel with ColorManagementSystem, and avoiding confusion with "unit systems" like Imperial Units and International Units.

Reply:
The registry can be used to keep a 1:1 relation between unit types and converters have have this done outside of shadergen (e.g. for software renderers). The UnitSystem which is mostly a construct for insertion of shader code As you may recall the registry and converters used to be in core but it was moved to shadergen as it's only used here.. It should still be possible to have conversion support without shader generation.  Ping @ashwinbhat for this as well.

Reply:
That makes sense, and thanks for the clarification.